### PR TITLE
Heartbeats for Sandbox-Next, ledger-on-sql, and ledger-on-memory.

### LINKED ISSUE #4580: Emit checkpoints from the KV SQL Ledger
Currently, the KV SQL ledger does not emit checkpoints, which are necessary for offsets and timeouts.

We will need to add a configurable heartbeat mechanism that supports at least:

- "wall-clock" time mode, which emits a checkpoint every second or so
- "static" time mode, which emits a checkpoint whenever the time is changed through the API

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
Heartbeat support for kvutils and its downstream usages, _ledger-on-memory_, _ledger-on-sql_, and _Sandbox-Next_.

This was done by turning kvutils' `LedgerRecord` into one implementation of a sealed trait, `LedgerEntry`; the other implementation is called `Heartbeat`. This means that the `stateUpdates` function now delivers a `Source[LedgerEntry]`, which is then picked up by the indexer.

When Sandbox-Next is run in static time (the current default), a heartbeat will be delivered whenever the clock is manually changed through the `TimeService.SetTime` GRPC function. In wall clock time, a heartbeat is sent every second.

Resolves #4580.

### Changelog

- **[Sandbox]** Fixed a bug in the command completions stream when running Sandbox in static time. Previously, upon updating the time, the old time was emitted on the completions stream. The new time is now emitted.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryLedgerReaderWriter.scala [hunk 1]
```diff
@@ -167,27 +135,34 @@ object InMemoryLedgerReaderWriter {
...
-        inMemoryState = new InMemoryState)
+        timeProvider,
+        dispatcher,
+        state,
+      )
     } yield readerWriter
+  }
 
-  // passing the Dispatcher and InMemoryState from the outside allows us to share
+  // passing the `dispatcher` and `inMemoryState` from the outside allows us to share
```
Comment (gerolf-da) [inline ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryLedgerReaderWriter.scala]:
```suggestion
  // passing the `dispatcher` and `state` from the outside allows us to share
```

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala [hunk 2]
```diff
@@ -131,15 +137,16 @@ object SqlLedgerReaderWriter {
...
-      val ledgerId = queries.updateOrRetrieveLedgerId(providedLedgerId)
-      if (initialLedgerId.exists(_ != ledgerId)) {
-        Future.failed(
-          new LedgerIdMismatchException(
-            domain.LedgerId(ledgerId),
-            domain.LedgerId(initialLedgerId.get),
-          ))
-      } else {
-        Future.successful(ledgerId)
+      Future.fromTry(queries.updateOrRetrieveLedgerId(providedLedgerId)).flatMap { ledgerId =>
```
Comment (gerolf-da) [inline ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: Considering the body of `flatMap` only constructs futures as `failed` and `successful` you can `flatMap` on the `Try` to avoid the context switch over the dispatcher.

File: ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala [hunk 3]
```diff
@@ -159,8 +161,8 @@ abstract class ScenarioLoadingITBase
             val events = responses.flatMap(extractEvents)
             val contractIds = events.map(_.contractId).toSet
 
-            // note how we skip #0, #2 and #5 because of the `pass`es in the scenario.
-            contractIds shouldBe Set("#4:0", "#7:2", "#3:0", "#7:1", "#1:0", "#7:0", "#6:0")
+            // note how we skip #0, #2, #5 and #7 because of the `pass`es in the scenario.
+            contractIds shouldBe Set("#1:0", "#3:0", "#4:0", "#6:0", "#8:0", "#8:1", "#8:2")
```
Comment (gerolf-da) [inline ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala]: What was the change for the increase of the contract id? A hearbeat after `#6:0`? Wonder how flaky this will be.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala [hunk 4]
```diff
@@ -38,19 +45,46 @@ object TimeServiceBackend {
...
-        .setCurrentTime(expectedTime, newTime)
-        .flatMap { success =>
-          if (success)
-            onTimeChange(expectedTime).map(_ => true)(DirectExecutionContext)
-          else Future.successful(false)
-        }(DirectExecutionContext)
+  private final class AkkaQueueBasedObservedTimeServiceBackend(delegate: TimeServiceBackend)(
+      implicit materializer: Materializer
+  ) extends ObservedTimeServiceBackend {
+    private val queues = mutable.Set[SourceQueueWithComplete[Instant]]()
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala]: What is the synchronization/locking mechanism that protects `queues` from inconsistent updates?

**Comment by gerolf-da** [comment]
I'd like @miklos-da or @dajmaki do way in on the changes for kvutils though.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala (hunk 4 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala]: Good question. Perhaps we should use a `concurrent.TrieMap` instead.

File: ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala (hunk 3 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala]: It's to do with the (somewhat accidental) bug fix, which was pushing out the wrong update to the time. Before, we were not recording the last `pass` in the ledger; now, we are.

In the test scenario, we have:

```daml
  pass (days 10)  -- #0
  party <- getParty "party"
  submit party (create Dummy with operator = party)  -- #1
  pass (seconds 10)  -- #2
  submit party (create DummyWithParam with operator = party)  -- #3
  submit party (create DummyFactory with operator = party)  -- #4
  pass (seconds 10)  -- #5
  submit party (create DummyContractFactory with operator = party)  -- #6
  pass (seconds 10)  -- #7
  other <- getParty "other"
  submitMustFail party (create Dummy with operator = other)
  return ()
```

Contracts `#8:0`–`#8.2` come from `dummyRequest.commands`.

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: Good point. Done!

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala (hunk 4 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala]: Done.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala [hunk 5]
```diff
@@ -40,7 +41,9 @@ class SubmissionValidator[LogResult](
     ) => LogEntryAndState,
     allocateLogEntryId: () => DamlLogEntryId,
     checkForMissingInputs: Boolean = false,
-)(implicit executionContext: ExecutionContext) {
+)(implicit executionContext: ExecutionContext, logCtx: LoggingContext) {
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: @miklos-da how public of an API is the SubmissionValidator? Will we use this directly or still wrap it with something? If it's the public API for integrations then isn't it a problem to require use of our internal logging here?

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: Good point. We can construct the logging context inside the `SubmissionValidator` if necessary to hide it from API consumers; this is basically the same as using slf4j directly.

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: `ValidatingCommitter` also exposes a dependency on `LoggingContext`. I'm going to remove it from the interface but still pass it through to the `SubmissionValidator.validateAndCommit` via a package-private method called `SubmissionValidator.validateAndCommitWithLoggingContext`. Does this make sense to do?

File: ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala (hunk 3 above)
Comment (gerolf-da) [reply ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala]: Ah sorry. the passing of time is driven by the scenario of course. So no time-based flakiness 👍. Sorry for the noise.

Comment (SamirTalwar-DA) [reply ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala]: No worries; it's a confusing test and it took me quite some time to understand it myself!

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala [hunk 6]
```diff
@@ -131,16 +137,17 @@ object SqlLedgerReaderWriter {
...
+        if (initialLedgerId.exists(_ != ledgerId)) {
+          Failure(
+            new LedgerIdMismatchException(
+              domain.LedgerId(ledgerId),
+              domain.LedgerId(initialLedgerId.get),
+            ))
+        } else {
+          Success(ledgerId)
+        }
+      })
```
Comment (miklos-da) [inline ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: The indentation here is a bit hard to follow. Can you make it easier to read, please? E.g., you could break the line after `Future.fromTry` or extract the construction of the `Try`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala (hunk 5 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: Yes, that seems like a good approach. `SubmissionValidator` is on the public API hence we should avoid requiring a `LoggingContext` in order to use it.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala [hunk 7]
```diff
@@ -38,19 +45,48 @@ object TimeServiceBackend {
...
+          case Success(true) =>
+            queues.keys.foreach(_.offer(newTime))
+        }(materializer.executionContext)
+
+    override def changes: ResourceOwner[Source[Instant, NotUsed]] =
+      new ResourceOwner[Source[Instant, NotUsed]] {
+        override def acquire()(
+            implicit executionContext: ExecutionContext
+        ): Resource[Source[Instant, NotUsed]] = {
+          Resource(Future {
```
Comment (miklos-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala]: The indentation is hard to follow here. Can you possibly break this apart into smaller bits to improve readability, please?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala [hunk 8]
```diff
@@ -298,13 +307,13 @@ final class SandboxServer(
...
                 submissionConfig = config.submissionConfig,
-                optTimeServiceBackend = timeServiceBackendO
-                  .map(TimeServiceBackend.withObserver(_, indexAndWriteService.publishHeartbeat)),
+                optTimeServiceBackend = observingTimeServiceBackend,
                 metrics = metrics,
                 healthChecks = healthChecks,
                 seedService = seedService,
               )(mat, esf, logCtx)
               .map(_.withServices(List(resetService(ledgerId, authorizer, executionContext)))),
+          // NOTE: Re-use the same port after reset.
```
Comment (miklos-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala]: What do you mean by this comment? The code says if you already have a port set then don't care about the port specified in the config. If you would like to imply that `currentPort` is the port saved before reset then please rename the parameter accordingly.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandboxnext/Runner.scala [hunk 9]
```diff
@@ -81,11 +81,13 @@ class Runner {
...
-    val timeServiceBackend = timeProviderType match {
+    val (timeServiceBackend, heartbeatMechanism) = timeProviderType match {
       case TimeProviderType.Static =>
-        Some(TimeServiceBackend.simple(Instant.EPOCH))
+        val backend = TimeServiceBackend.observing(TimeServiceBackend.simple(Instant.EPOCH))
+        (Some(backend), backend.changes)
       case TimeProviderType.WallClock =>
-        None
+        val clock = Clock.systemUTC()
+        (None, new RegularHeartbeat(clock, 1.second))
```
Comment (miklos-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandboxnext/Runner.scala]: Shouldn't this be configurable or at least specified by a constant?

File: ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryState.scala [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      action(log, state)
+    } finally {
+      lockCurrentState.release()
+    }
+  }
+
+  def withFutureWriteLock[A](action: (MutableLog, MutableState) => Future[A])(
+      implicit executionContext: ExecutionContext
+  ): Future[A] = {
+    lockCurrentState.acquire()
```
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryState.scala]: Yes. If we acquired the lock afterwards, what would be the point?

AFAIK, this has exactly the same semantics as:

```
    Future.successful(lockCurrentState.acquire())
      .flatMap(_ => action(log, state))
      .andThen { case _ => lockCurrentState.release() }

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala (hunk 8 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala]: This comment was just moved from the code above because it no longer fit; I didn't change it, and I'd rather not change it in this PR.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandboxnext/Runner.scala (hunk 9 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandboxnext/Runner.scala]: Good call. I'll pull out a constant.

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala [hunk 11]
```diff
@@ -131,16 +137,17 @@ object SqlLedgerReaderWriter {
...
+      Future.fromTry(queries.updateOrRetrieveLedgerId(providedLedgerId).flatMap { ledgerId =>
+        if (initialLedgerId.exists(_ != ledgerId)) {
+          Failure(
+            new LedgerIdMismatchException(
+              domain.LedgerId(ledgerId),
+              domain.LedgerId(initialLedgerId.get),
+            ))
+        } else {
+          Success(ledgerId)
+        }
```
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: Solid point. Should be easier now.

File: ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryState.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (dajmaki) [reply ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryState.scala]: Would it make sense to do this inside a `Future { }` block so that acquiring the lock happens in the right execution context?

[the remark this answers is not in the document]
Comment (miklos-da) [reply ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryState.scala]: That's fine, just wanted to check if you are good with immediately blocking the current thread when you call `withFutureWriteLock`.
Of course, didn't want to question whether you need to lock before or after `action` (edited comment).

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala (hunk 8 above)
Comment (miklos-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala]: OK -- I'm just pointing out this comment is rather confusing than helping.

File: ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryState.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryState.scala]: Doesn't really matter; this is not a lock that's bound to a thread, it's a simple semaphore which can be acquired and released from anywhere. The developer is responsible for ensuring that it's not released unless it's acquired.

@miklos-da, you wrote this code, I just moved it. 😉
